### ISSUE TITLE
[WIP] Correctify Unicode-related errors and omissions in "Lexical structure"

### ISSUE THREAD

**Issue by srutzky** [dotnet/csharplang#2672]
{ This is a work in progress.. I will be updating this to add more content and examples over the next hour }
==

Background
--

Large portions of the Unicode info in the "[Lexical structure](https://github.com/dotnet/csharplang/blob/master/spec/lexical-structure.md)" documentation is either missing, incorrect, or misleading. The main area of concern is with Unicode escape sequences, especially related to supplementary characters. And, it has been this way since the very beginning (almost 20 years now?). It has not helped that there was no example of the "\U" escape sequence anywhere, and the only description was in the ["String Escape Sequences" section of the C# Strings documentation](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/strings/#string-escape-sequences) which said, "for surrogate pairs.", which is not exactly correct, but definitely misleading. That wording, coupled with the lack of any examples to guide interpretation, leads one to think that you specify the two surrogate code points as there are 8 hex digits available and that's what is required for two UTF-16 surrogate code points.

Cha cha cha changes to this page
--

1. Broke `unicode_escape_sequence` token into two tokens: `unicode_bmp_escape_sequence` and `unicode_supplementary_escape_sequence`. This not only accurately reflects reality, but it also helps distinguish in the ANTLR grammar (not just in text comments) the set of Unicode characters that can be used in character literals and identifiers vs those that are invalid for those two cases.

2. Specify in ANTLR grammar the constrained usage of "\U" that creates BMP characters and is thus equivalent to "\u" where "\u" is valid to use.

3. Specify in ANTLR grammar how to create supplementary characters using both "\u" to specify an actual surrogate pair, and "\U" to specify a Unicode code point / UTF-32 code unit (they are synonymous). Along with the updates I recently made to the ["Escape Sequences" section of the C# Strings documentation](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/strings/#string-escape-sequences) to show usable range and an example, this should prevent any more misinterpreting of "\U".

4. Introduce the concept of "Supplementary Characters". That term has never been used in this document, which I believe has lead some to mistakenly use the term "surrogate pairs" (which was mentioned here only once, though not incorrectly) to mean "supplementary characters". Surrogate Pairs is a UTF-16 encoding-specific construct. It does not refer to the set of characters, but instead to how they are physically encoded, but only in UTF-16. Neither UTF-8 nor UTF-32 have surrogate pairs, yet they both certainly support all supplementary characters.

5.  Clean-up some inconsistent terminology and formatting:
    1. "code point" vs "value" 
    2. U+1234 vs `U+1234`
    3. U+1234 vs 0x1234

6. Provide an accurate and thorough description of BMP range/characters vs Supplementary range/characters, what constitutes a valid surrogate pair, that surrogate pairs are an encoding construct specific to UTF-16, and how all of that relates to `char` vs `string` literals and usage of "`\u`" and "`\U`".

7. Clarify that only BMP characters (direct or via `unicode_bmp_escape_sequence` ) can be used in identifiers. Previously, the ANTLR grammar did not prevent supplementary characters from being used, and there was not text description stating that they couldn't be used. However, they cannot be used (at least not in LINQPad 5 or VS 2015 -- related request: https://github.com/dotnet/csharplang/issues/1742 -- though "gmcs 4.6.2" via Mono does seem to allow them, possibly due to incomplete supplementary character definition here).

8. Added new token for "`hex_digit_except_zero`" to be used in the new "`unicode_supplementary_escape_sequence`" token, else the definition for "unicode_supplementary_escape_sequence" would have been rather messy / harder to read.

9. Adjusted definition of "`hex_digit`" to use new "`hex_digit_except_zero`" token. All previous uses of "`hex_digit`" are unaffected.

10. Some English grammatical improvements to "Note:" at beginning of **Character literals** section.

11. Updated ANTLR grammar for **Character literals** to specify only BMP characters can be used. Previously the grammar did not prevent supplementary characters from being used, and it was only noted in the text description.

12. Removed the "A Unicode character escape sequence ... must be in the range U+0000 to U+FFFF." sentence from the **Character literals** section as it is now doubly redundant, given the sentence about "code points above U+FFFF will error", _and_ the ANTLR grammar finally restriction the values to only the allowable range.

Done?
--
No, not done. There are still some areas that need correction, but they are out of scope for now, and can be handled separately (and might require additional discussion).

---

For more info on this project, please see:
     [Unicode Escape Sequences Across Various Languages and Platforms (including Supplementary Characters)](https://sqlquantumleap.com/2018/09/28/native-utf-8-support-in-sql-server-2019-savior-false-prophet-or-both/#csharp)

**Comment by ufcpp** [comment]
partially related to https://github.com/dotnet/roslyn/issues/13474
It's a known issue that Roslyn does not recognize surrogate pairs in identifiers.

**Comment by BillWagner** [comment]
moving to dotnet/csharpstandard

Note that @gafter refers to the [associated PR](https://github.com/dotnet/csharplang/pull/2675) in #155 (I suggest we close that issue as a duplicate, as the discussion is primarily in this issue).
